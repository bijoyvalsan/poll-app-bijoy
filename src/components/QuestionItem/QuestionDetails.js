import React from "react";
import PropTypes from "prop-types";

import SaveVote from "../Button";

import {
  StyledContainer,
  StyledContainerItem,
  StyledDeailsContainer,
  StyledDeailsColumn,
  StyledDeailsRow,
  StyledSubTitle,
  StyledBox
} from "./QuestionDetails.styled";

const QuestionDetails = ({ question, onVote }) => (
  <StyledBox>
    <StyledSubTitle>Question: {question.question}</StyledSubTitle>
    <StyledContainer>
      {question.choices.map((c, i) => (
        <StyledContainerItem key={c.url}>
          <StyledDeailsContainer>
            <StyledDeailsColumn>
              <StyledDeailsRow>
                Choice {i + 1}: {c.choice}
              </StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <StyledDeailsRow>Votes: {c.votes}</StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <StyledDeailsRow>Percentage: {c.percentage}%</StyledDeailsRow>
            </StyledDeailsColumn>
            <StyledDeailsColumn>
              <SaveVote.Button
                data-testid={`save-vote-${i}`}
                onClick={onVote(c.url)}
              >
                Save Vote
              </SaveVote.Button>
            </StyledDeailsColumn>
          </StyledDeailsContainer>
        </StyledContainerItem>
      ))}
    </StyledContainer>
  </StyledBox>
);

QuestionDetails.defaultProps = {
  onVote: () => {}
};

QuestionDetails.propTypes = {
  onVote: PropTypes.func,
  question: PropTypes.shape({
    id: PropTypes.string,
    question: PropTypes.string,
    choices: PropTypes.arrayOf(
      PropTypes.shape({
        choice: PropTypes.string.isRequired,
        url: PropTypes.string.isRequired,
        votes: PropTypes.number.isRequired,
        percentage: PropTypes.number.isRequired
      })
    )
  })
};

export default QuestionDetails;
