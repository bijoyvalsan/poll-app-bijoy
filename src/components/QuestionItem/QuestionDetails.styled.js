import styled from "styled-components";

export const StyledContainer = styled.div`
  background-color: white;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  box-shadow: rgba(12, 15, 20, 0.02) 0px 0px 0px 1px,
    rgba(12, 15, 20, 0.06) 0px 0px 1px 0px,
    rgba(12, 15, 20, 0.06) 0px 2px 2px 0px;
  padding: 0px;
  margin-bottom: 16px;
`;

export const StyledContainerItem = styled.div`
  align-items: center;
  position: relative;
  border-bottom: 1px solid #d8dde1;
  padding: 16px;
  &:nth-child(even) {
    background-color: #f8f8f8;
  }
`;

export const StyledDeailsContainer = styled.div`
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
`;

export const StyledDeailsColumn = styled.div`
  display: flex;
  align-items: center;
  justify-content: start;
  flex: 1;
  min-width: 100px;
  box-sizing: border-box;
`;

export const StyledDeailsRow = styled.div`
  margin-right: 16px;
`;

export const StyledSubTitle = styled.p`
  font-size: 20px;
`;

export const StyledBox = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
`;
