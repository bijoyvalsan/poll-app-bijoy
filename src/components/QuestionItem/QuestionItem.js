import React from "react";
import PropTypes from "prop-types";

import {
  StyledContainer,
  StyledQuestionTitle,
  StyledQuestionPublishDate,
  StyledQuestionChoices
} from "./QuestionItem.styled";

const QuestionItem = ({ questionItem, onClick }) => (
  <StyledContainer pointer onClick={onClick}>
    <StyledQuestionTitle>{questionItem.question}</StyledQuestionTitle>
    <StyledQuestionPublishDate>
      Published: {questionItem.publishedAt}
    </StyledQuestionPublishDate>
    <StyledQuestionChoices>
      No. Choices: {questionItem.noOfChoices}
    </StyledQuestionChoices>
  </StyledContainer>
);

QuestionItem.defaultProps = {
  onClick: () => {}
};

QuestionItem.propTypes = {
  onClick: PropTypes.func,
  questionItem: PropTypes.shape({
    id: PropTypes.string.isRequired,
    question: PropTypes.string.isRequired,
    publishedAt: PropTypes.string.isRequired,
    noOfChoices: PropTypes.number.isRequired
  }).isRequired
};

export default QuestionItem;
