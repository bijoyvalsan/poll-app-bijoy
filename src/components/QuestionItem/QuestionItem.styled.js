import styled, { keyframes, css } from "styled-components";
import { device } from "../../theme";

const pulse = keyframes`
  0% {
    background-position: 0% 0%;
  }
  100% {
    background-position: -135% 0%;
  }
`;

export const StyledContainer = styled.div`
  background-color: white;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  box-shadow: 0px 3px 3px -2px rgb(0 0 0 / 20%),
    0px 3px 4px 0px rgb(0 0 0 / 14%), 0px 1px 8px 0px rgb(0 0 0 / 12%);
  display: flex;
  flex-direction: column;
  margin: 12px;
  padding: 16px;
  max-width: 300px;
  min-width: 300px;
  box-sizing: border-box;
  cursor: ${({ pointer }) => (pointer ? "pointer" : "inherit")};
  @media ${device.mobile} {
    max-width: 250px;
    min-width: 250px;
  }
`;

const StyledItem = styled.div`
  display: inline-block;
  height: 100%;
  width: 100%;
  background: ${props =>
    props.translucent
      ? css`linear-gradient(-90deg, #DEDEDE 0%, #F8F8F8 50%, #DEDEDE 100%)`
      : css`linear-gradient(-90deg, #F1F1F1 0%, #F8F8F8 50%, #F1F1F1 100%)`};
  background-size: 400% 400%;
  animation: ${pulse} 2s ease-in-out 0s infinite;
`;

export const StyledLoader = styled(StyledItem)`
  width: ${({ width }) => width || "70px"};
  height: ${({ height }) => height || "20px"};
  border-radius: ${({ radius }) => radius || "20px"};
  margin: 8px;
  &::before {
    ${() =>
      css`
        content: "\00a0";
      `}
  }
`;

export const StyledQuestionTitle = styled.p`
  font-size: 18px;
  line-height: 20px;
  font-weight: 500;
`;

export const StyledQuestionPublishDate = styled.span`
  opacity: 0.6;
`;

export const StyledQuestionChoices = styled.span`
  opacity: 0.6;
`;
