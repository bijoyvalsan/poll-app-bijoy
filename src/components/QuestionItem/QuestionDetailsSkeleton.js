import React from "react";

import { StyledLoader } from "./QuestionItem.styled";
import {
  StyledDeailsContainer,
  StyledDeailsColumn,
  StyledDeailsRow,
  StyledBox,
  StyledContainer,
  StyledContainerItem
} from "./QuestionDetails.styled";

const QuestionDetailsSkeleton = () => (
  <StyledBox>
    <div>
      <StyledLoader width="100px" />
    </div>
    <StyledContainer>
      <StyledContainerItem>
        <StyledDeailsContainer>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
        </StyledDeailsContainer>
      </StyledContainerItem>
      <StyledContainerItem>
        <StyledDeailsContainer>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
        </StyledDeailsContainer>
      </StyledContainerItem>
      <StyledContainerItem>
        <StyledDeailsContainer>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
          <StyledDeailsColumn>
            <StyledDeailsRow>
              <StyledLoader translucent width="100px" />
            </StyledDeailsRow>
          </StyledDeailsColumn>
        </StyledDeailsContainer>
      </StyledContainerItem>
    </StyledContainer>
  </StyledBox>
);

export default QuestionDetailsSkeleton;
