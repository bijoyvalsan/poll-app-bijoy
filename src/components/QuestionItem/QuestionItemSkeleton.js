import React from "react";

import { StyledContainer, StyledLoader } from "./QuestionItem.styled";

const QuestionItemSkeleton = () => (
  <StyledContainer>
    <StyledLoader width="60%" />
    <StyledLoader translucent width="90%" />
    <StyledLoader translucent width="90%" />
  </StyledContainer>
);

export default QuestionItemSkeleton;
