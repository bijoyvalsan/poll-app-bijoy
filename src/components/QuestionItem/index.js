import QuestionItem from "./QuestionItem";
import QuestionItemSkeleton from "./QuestionItemSkeleton";
import QuestionDetails from "./QuestionDetails";
import QuestionDetailsSkeleton from "./QuestionDetailsSkeleton";

const Question = {
  QuestionItem,
  QuestionDetails,
  QuestionItemSkeleton,
  QuestionDetailsSkeleton
};

export default Question;
