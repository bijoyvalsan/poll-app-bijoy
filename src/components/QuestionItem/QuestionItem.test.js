import React from "react";
import { render } from "@testing-library/react";

import QuestionItem from "./QuestionItem";
import QuestionItemSkeleton from "./QuestionItemSkeleton";

describe("QuestionItem Component", () => {
  it("should render with default QuestionItem styles", () => {
    const mockData = {
      id: "1",
      question: "Who am i ?",
      publishedAt: "12/12/1991 12:00",
      noOfChoices: 3
    };
    const { asFragment } = render(<QuestionItem questionItem={mockData} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("QuestionItemSkeleton Component", () => {
  it("should render with default QuestionItemSkeleton styles", () => {
    const { asFragment } = render(<QuestionItemSkeleton />);
    expect(asFragment()).toMatchSnapshot();
  });
});
