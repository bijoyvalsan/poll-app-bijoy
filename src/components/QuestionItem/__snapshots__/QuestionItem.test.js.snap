// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionItem Component should render with default QuestionItem styles 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM goBlJX"
  >
    <p
      class="sc-hKFxyN kSMtjo"
    >
      Who am i ?
    </p>
    <span
      class="sc-eCApnc PowZy"
    >
      Published: 12/12/1991 12:00
    </span>
    <span
      class="sc-jSFjdj jwWkJz"
    >
      No. Choices: 3
    </span>
  </div>
</DocumentFragment>
`;

exports[`QuestionItemSkeleton Component should render with default QuestionItemSkeleton styles 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnxRM czhEWd"
  >
    <div
      class="sc-gtsrHT sc-dlnjwi kPrUCA biuPjS"
      width="60%"
    />
    <div
      class="sc-gtsrHT sc-dlnjwi igZhXg fMRwIx"
      width="90%"
    />
    <div
      class="sc-gtsrHT sc-dlnjwi igZhXg fMRwIx"
      width="90%"
    />
  </div>
</DocumentFragment>
`;
